src/rust/inetstack/protocols/layer3/icmpv4/header.rs

```diff
@@ -70,25 +66,26 @@ impl Icmpv4Header {
         let icmpv4_type: Icmpv4Type2 = Icmpv4Type2::parse(type_byte, rest_of_header)?;
 
         buf.adjust(ICMPV4_HEADER_SIZE)?;
-        Ok((
-            Self {
-                protocol: icmpv4_type,
-                code,
-            },
-            buf,
-        ))
+        Ok(Self {
+            protocol: icmpv4_type,
+            code,
+        })
     }
 
-    pub fn serialize(&self, buf: &mut [u8], data: &[u8]) {
-        let buf: &mut [u8; ICMPV4_HEADER_SIZE] = (&mut buf[..ICMPV4_HEADER_SIZE]).try_into().unwrap();
+    /// Serializes and prepends the ICMP header into the packet in [buf]. This function assumes that the packet has
+    /// sufficient headroom to fit the ICMP header.
+    pub fn serialize_and_attach(&self, buf: &mut DemiBuffer) {
+        buf.prepend(ICMPV4_HEADER_SIZE).expect("Should have headroom");
+
         let (type_byte, rest_of_header) = self.protocol.serialize();
         buf[0] = type_byte;
         buf[1] = self.code;
         // Skip the checksum for now.
         buf[2] = 0;
         buf[3] = 0;
         buf[4..8].copy_from_slice(&rest_of_header[..]);
-        let checksum: u16 = Self::compute_checksum(buf, data);
+        let (hdr_buf, payload): (&[u8], &[u8]) = buf[..].split_at(ICMPV4_HEADER_SIZE as usize);
+        let checksum: u16 = Self::compute_checksum(hdr_buf.try_into().unwrap(), payload);
```

Comment:
Not sure if this try here is required.

Reply:
What would be the alternative? Just use into()? I guess that would be a compile-time check...? That would be better IMO.

Reply:
yes, I think we could use just an into here (which I think would happen automatically)